### PR TITLE
[Eval] Add evaluation script with full steps accuracy

### PR DESCRIPTION
**Pull request by ucinereo**
Added new metric to log whether the full sequence is predicted correctly (c.f. https://arxiv.org/abs/2405.17399v1)

## Evaluation Script
New file `eval.py` which takes a config file (e.g. `eval_config.yaml`) as parameter:
```yaml
outfile: evaluation_results.txt
min_digit: 3
max_digit: 100
digits: [3, 4, 5, 6, 7, 8, 9, 10, 20, 50, 100]
samples: 1000
metrics:
  - step_accuracy
  - full_sample_accuracy

models:
  # [DisplayName, PathToOutputFolder, CheckpointNumber]
  - ["SplitHead ($\\beta=7$)", "outputs/split_head_1_7/2026-01-12/17-11-14", 200]
  - ["SplitHead ($\\beta=4$)", "outputs/split_head_1_4/2026-01-12/06-47-42", 200]
```
For each `metric` it prints a table per metric and stores all tables in latex format in `outfile`.

### CODE REVIEW

File: eval.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+            m_ckpt = f"ckpt_{m_ckpt}.pth"
+
+            results[m_name] = {}
+
+            logger.info(f"Evaluating Model: {m_name}")
+
+            for d in digit_range:
+                with PrintSuppressor():
+                    trainer = load_trainer(m_path, m_ckpt, n_samples, d)
```
Comment (gserifi) [inline eval.py:125]: It would be cleaner to decouple `d` from load_trainer, as it takes a bit of time to load the checkpoints, which would be independent of `d` though. However, then you'd have to manually instantiate a dataset object.

File: train.py [hunk 2]
```diff
@@ -290,6 +290,67 @@ def val_epoch(self, epoch: int) -> tuple[float, float]:
...
+                (1, val_dataset.h, val_dataset.w), dtype=torch.long
+            ) * val_dataset.token_to_idx.index("\n")
+
+            # Parse the first step string
+            s = gt_steps[0]
+            for i in range(val_dataset.h):
+                for j in range(val_dataset.w - 1):  # \n already filled
+                    current_grid[0, i, j] = val_dataset.token_to_idx.index(s.split("\n")[i][j])
+
+            current_grid = current_grid.to(device)
```
Comment (gserifi) [inline train.py]: better to use `self.fabric.to_device(current_grid)`

File: train.py [hunk 3]
```diff
@@ -290,6 +290,67 @@ def val_epoch(self, epoch: int) -> tuple[float, float]:
...
+
+            for _ in range(steps_to_rollout):
+                with torch.no_grad():
+                    _, _, y_pred, _, _, _, _, _, _ = self.forward((current_grid, current_grid))
+
+                    if isinstance(y_pred, tuple):
+                        y_pred_mapped = tuple(y_p for y_p in y_pred)
```
Comment (gserifi) [inline train.py]: What's the purpose of this?

**Review by gserifi — CHANGES_REQUESTED** [review]
Looks good, we can add some extra features later, like automatic checkpoint selection etc.
I think fixing the unnecessary model loading would be nice, but it would probably require quite a few modifications to code in some other places. Feel free to ignore that if you feel like it's not worth it.

File: eval.py (hunk 1 above)
Comment (ucinereo) [reply eval.py:125]: compared to the evaluation of 1000 steps, the checkpoint loading is negligible

File: train.py (hunk 3 above)
Comment (ucinereo) [reply train.py]: just wrongly copied from `val_k_epoch`. will delete that

File: eval.py (hunk 1 above)
Comment (gserifi) [reply eval.py:125]: Ok, then let's not worry about it

**Review by gserifi — APPROVED** [review]
